src/codegen/dispatch/substrate.rs

```diff
@@ -0,0 +1,388 @@
+
+        self.add(Instr::Call {
+            res: returns,
+            call: InternalCallTy::Static { cfg_no: msg_no },
+            args,
+            return_tys,
+        });
+
+        if cfg.returns.is_empty() {
+            let data_len = Expression::NumberLiteral(Codegen, Uint(32), 0.into());
+            let data = Expression::AllocDynamicBytes(
+                Codegen,
+                Type::DynamicBytes,
+                data_len.clone().into(),
+                None,
+            );
+            self.add(Instr::ReturnData { data, data_len })
+        } else {
+            let (data, data_len) = abi_encode(
+                &Codegen,
+                returns_expr,
+                self.ns,
+                &mut self.vartab,
+                &mut self.cfg,
+                false,
+            );
+            self.add(Instr::ReturnData { data, data_len });
+        }
+
+        case_bb
+    }
+
+    /// Insert a trap into the cfg, if the message `msg_no` is not payable but received value anyways.
+    /// Constructors always receive endowment.
+    fn abort_if_value_transfer(&mut self, msg_no: usize) {
+        if !self.all_cfg[msg_no].nonpayable || self.all_cfg[msg_no].ty == FunctionTy::Constructor {
+            return;
+        }
+
+        let value_ty = Uint(self.ns.value_length as u16 * 8);
+        let true_block = self.cfg.new_basic_block(format!("msg_{msg_no}_has_value"));
+        let false_block = self.cfg.new_basic_block(format!("msg_{msg_no}_no_value"));
+        self.add(Instr::BranchCond {
+            cond: Expression::More {
+                loc: Codegen,
+                signed: false,
+                left: Expression::Variable(Codegen, value_ty.clone(), self.value).into(),
+                right: Expression::NumberLiteral(Codegen, value_ty, 0.into()).into(),
+            },
+            true_block,
+            false_block,
+        });
+
+        self.cfg.set_basic_block(true_block);
+        let function_name = self.all_cfg[msg_no].name.split("::").last().unwrap();
+        log_runtime_error(
+            self.opt.log_runtime_errors,
+            &format!("runtime_error: non payable function {function_name} received value"),
+            Codegen,
+            &mut self.cfg,
+            &mut self.vartab,
+            self.ns,
+        );
+        self.add(Instr::AssertFailure { encoded_args: None });
+
+        self.cfg.set_basic_block(false_block);
+    }
+
+    /// Build calls to fallback or receive functions (if they are present in the contract).
+    fn fallback_or_receive(&mut self) {
+        let fb_recv = self
+            .all_cfg
+            .iter()
+            .enumerate()
+            .fold([None, None], |mut acc, (no, cfg)| {
+                match cfg.ty {
+                    FunctionTy::Fallback if cfg.public => acc[0] = Some(no),
+                    FunctionTy::Receive if cfg.public => acc[1] = Some(no),
+                    _ => {}
+                }
+                acc
+            });
```

Comment:
I like your single-scan of all_cfg to find the fallback/receive but it's not easy to read.

how about starting with `let (fallback_cfg, receive_cfg) = self.all_cfg...`